File: POLIMORFISMO/Ejercicio 3/codigo/codigo.py
#3.Sean las siguientes clases que hacen referencia a diferentes tipos de figura:
#   a)Instanciar 1 Cuadrado, 1 Rectángulo y 1 Círculo
#   b)Implementar un método mostrar() en cada clase para imprimir sus dimensiones y área.
#   c)Sobrecargar el método Área para mostrar el área de todas las figuras.
import math
class Figura:
    def mostrar(self):
        pass
    
    def area(self):
        pass
class Cuadrado(Figura):
    def __init__(self, lado):
        self.lado = lado
    
    def mostrar(self):
        print(f"Cuadrado - Lado: {self.lado}, Área: {self.area()}")
    
    def area(self):
        return self.lado ** 2
class Rectangulo(Figura):
    def __init__(self, base, altura):
        self.base = base
        self.altura = altura
    
    def mostrar(self):
        print(f"Rectángulo - Base: {self.base}, Altura: {self.altura}, Área: {self.area()}")
    
    def area(self):
        return self.base * self.altura
class Circulo(Figura):
    def __init__(self, radio):
        self.radio = radio
    
    def mostrar(self):
        print(f"Círculo - Radio: {self.radio}, Área: {self.area():.2f}")
    
    def area(self):
        return math.pi * (self.radio ** 2)

def mostrarareas(*figuras):
    for figura in figuras:
        print(f"Área de {figura.__class__.__name__}: {figura.area():.2f}")

cuadrado = Cuadrado(5)
rectangulo = Rectangulo(4, 6)
circulo = Circulo(3)

cuadrado.mostrar()
rectangulo.mostrar()
circulo.mostrar()

print("\nMostrando áreas de todas las figuras:")
mostrarareas(cuadrado, rectangulo, circulo)

File: POLIMORFISMO/Ejercicio 3/codigo/codigo.java
//3.Sean las siguientes clases que hacen referencia a diferentes tipos de figura:
//  a)Instanciar 1 Cuadrado, 1 Rectángulo y 1 Círculo
//  b)Implementar un método mostrar() en cada clase para imprimir sus dimensiones y área.
//  c)Sobrecargar el método Área para mostrar el área de todas las figuras.
import java.util.ArrayList;
import java.util.List;
abstract class Figura {
    public abstract void mostrar();
    public abstract double area();
}
class Cuadrado extends Figura {
    private double lado;
    public Cuadrado(double lado) {
        this.lado = lado;
    }
    @Override
    public void mostrar() {
        System.out.printf("Cuadrado - Lado: %.2f, Área: %.2f%n", lado, area());
    }
    @Override
    public double area() {
        return lado * lado;
    }
}
class Rectangulo extends Figura {
    private double base;
    private double altura;
    
    public Rectangulo(double base, double altura) {
        this.base = base;
        this.altura = altura;
    }
    @Override
    public void mostrar() {
        System.out.printf("Rectángulo - Base: %.2f, Altura: %.2f, Área: %.2f%n", 
                         base, altura, area());
    }
    
    @Override
    public double area() {
        return base * altura;
    }
}
class Circulo extends Figura {
    private double radio;
    private static final double PI = Math.PI;
    public Circulo(double radio) {
        this.radio = radio;
    }
    @Override
    public void mostrar() {
        System.out.printf("Círculo - Radio: %.2f, Área: %.2f%n", radio, area());
    }
    @Override
    public double area() {
        return PI * radio * radio;
    }
}

public class Main {
    public static void mostrarAreas(Figura... figuras) {
        for (Figura figura : figuras) {
            System.out.printf("Área de %s: %.2f%n", 
                            figura.getClass().getSimpleName(), 
                            figura.area());
        }
    }
    public static void main(String[] args) {
        Cuadrado cuadrado = new Cuadrado(5);
        Rectangulo rectangulo = new Rectangulo(4, 6);
        Circulo circulo = new Circulo(3);
        cuadrado.mostrar();
        rectangulo.mostrar();
        circulo.mostrar();
        System.out.println("\nMostrando áreas de todas las figuras:");
        mostrarAreas(cuadrado, rectangulo, circulo);
    }
}